src/server/lib/okta/idx/interact.ts

```diff
@@ -0,0 +1,168 @@
+	res: ResponseWithRequestState,
+	{
+		confirmationPagePath,
+		closeExistingSession,
+		doNotSetLastAccessCookie = false,
+		extraData,
+	}: Pick<
+		PerformAuthorizationCodeFlowOptions,
+		| 'confirmationPagePath'
+		| 'closeExistingSession'
+		| 'doNotSetLastAccessCookie'
+		| 'extraData'
+	>,
+): Promise<[InteractResponse, AuthorizationState]> => {
+	try {
+		// determines if the current session, should it exist, be closed
+		if (closeExistingSession) {
+			// Okta Identity Engine session cookie is called `idx`
+			const oktaIdentityEngineSessionCookieId: string | undefined =
+				req.cookies.idx;
+			if (oktaIdentityEngineSessionCookieId) {
+				await closeCurrentSession({
+					idx: oktaIdentityEngineSessionCookieId,
+				});
+			}
+		}
+
+		// generate the code verifier and code challenge for PKCE
+		// see https://www.oauth.com/oauth2-servers/pkce/authorization-request/
+		// the `code_verifier` is a cryptographically random string
+		// the `code_challenge` is the `code_verifier` hashed with SHA-256 and then base64url encoded
+		const codeVerifier = generators.codeVerifier(64);
+		const codeChallenge = generators.codeChallenge(codeVerifier);
+
+		// generate and store a "state"
+		// as a http only, secure, signed session cookie
+		// which is a json object that contains a stateParam and the query params
+		// the stateParam is used to protect against csrf
+		// we also set additional data in the state cookie that we need to persist
+		// between the authorization request and the callback
+		const authState = generateAuthorizationState(
+			getPersistableQueryParams(res.locals.queryParams),
+			confirmationPagePath,
+			doNotSetLastAccessCookie,
+			{
+				appPrefix: extraData?.appPrefix,
+				encryptedRegistrationConsents: extraData?.encryptedRegistrationConsents,
+				codeVerifier,
+			},
+		);
+		setAuthorizationStateCookie(authState, res);
+
+		// set up the authorization parameters to send to the /interact endpoint
+		// in order to start the interaction code flow
+		// see https://developer.okta.com/docs/concepts/interaction-code/
+		const authorizationParams: AuthorizationParameters = {
+			// the client_id is the id of the client application
+			client_id: okta.clientId,
+			// only use the interaction code flow callback uri, everything else is handled by standard authorization code flow
+			redirect_uri: ProfileOpenIdClientRedirectUris.INTERACTION_CODE,
+			// only authentication scopes for interaction code flow, everything else is handles by standard authorization code flow
+			scope: scopesForAuthentication.join(' '),
+			// we send the generated code challenge
+			code_challenge: codeChallenge,
+			// we use the S256 code challenge method, which is the only method supported
+			code_challenge_method: 'S256',
+			// we send the generated stateParam as the state parameter
+			state: authState.stateParam,
+		};
+
+		const authorizationSearchParams = new URLSearchParams(
+			authorizationParams as Record<string, string>,
```

Comment:
Is this typecast really necessary here?

Reply:
Unfortunately yes, as the `AuthorizationParams` type doesn't overlap with the URLSearchParams input type, even though it's mostly `Record<string, string>`.

```
Argument of type 'AuthorizationParameters' is not assignable to parameter of type 'string | Record<string, string> | string[][] | URLSearchParams | undefined'.
Type 'AuthorizationParameters' is missing the following properties from type 'string[][]': length, pop, push, concat, and 29 more.ts(2345)
⚠ Error (TS2345)  | 

Argument of type 
AuthorizationParameters
 is not assignable to parameter of type:
| string
| Record<string, string>
| string[][]
| URLSearchParams
| undefined
.
  | Type  AuthorizationParameters  is missing the following properties from type  string[][]  :lengthpoppushconcatand 29 more. -- | --
  ```